### PR TITLE
[gpt2] Fix GPT-2 generation when using batchsize >1

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
@chinnadhurai recently observed that when using batchsize > 1, the generations of GPT2 don't match those when using batchsize = 1, resulting in a significant loss of performance.

Upon inspection, the bug was around how positional embeddings were being treated with padding. HF's GPT-2 implementation defaults to using `arange`, rather than counting non-padding tokens, which caused incongruences in the positional embeddings used for generations with unequal number of context tokens. This was particularly cumbersome after the first generation, and exacerbated by our beam search's speedup which only computes output softmax on the final timestep.

Fortunately, HF's implementation allows for manually specifying both the padding and position embeddings as input. This diff reworks our wrapper around HF's code so that these things are computed on our side. The result is actually a cleaner implementation, in my humble opinion.

**Testing steps**
Trained a model for one epoch on [redacted], then tested with:
- bs 1
- bs 2, 4, 8, 16
- dynamic batching bs 8
- multiprocessing eval with bs 8

Final results were _slightly different_ (fourth decimal place), but in the final ballpark.

Also added new CI for GPT2 which tests generation with different batchsizes.

### CODE REVIEW

**Comment by stephenroller** [comment]
I've also slipped in some slightly better initialization to the special tokens. There's little improvement after as few as 200 steps, but it keeps the first couple prints from being out of control.

File: parlai/agents/hugging_face/gpt2.py [hunk 1]
```diff
@@ -35,12 +34,21 @@ def __init__(self, opt, dict):
...
-            self.add_start_token = opt["add_start_token"]
+            with torch.no_grad():
+                # first reduce the random jitter of the initialization
+                self.transformer.wte.weight[size_before:] *= 0.1
+                # next center it on the endoftext token
+                self.transformer.wte.weight[
+                    size_before:
+                ] += self.transformer.wte.weight[size_before - 1].unsqueeze(0)
+
+        self.add_start_token = opt["add_start_token"]
```
Comment (emilydinan) [inline parlai/agents/hugging_face/gpt2.py:48]: should this go above? will the transformer need to be resized depending on this value?

File: parlai/agents/hugging_face/gpt2.py [hunk 2]
```diff
@@ -253,11 +252,27 @@ def add_cmdline_args(cls, argparser):
         return agent
 
     def __init__(self, opt, shared=None):
-        if not opt["add_special_tokens"] and opt["batchsize"] > 1:
+        if not opt["add_special_tokens"] and opt.get('batchsize', 1) > 1:
+            # *** STOP ***
```
Comment (emilydinan) [inline parlai/agents/hugging_face/gpt2.py:256]: lmao

File: tests/nightly/gpu/test_gpt2.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import unittest
+from parlai.core.agents import create_agent
+
+
+class TestGpt2(unittest.TestCase):
```
Comment (emilydinan) [inline tests/nightly/gpu/test_gpt2.py:11]: nice

File: parlai/agents/hugging_face/gpt2.py [hunk 4]
```diff
@@ -66,33 +74,62 @@ def _init_from_pretrained(self, opt):
...
             else:
                 # forced decoding: concatenate the context
                 # with the labels
-                model_input, _ = concat_without_padding(
-                    encoder_state, input, use_cuda=self.use_cuda, null_idx=self.null_idx
-                )
-                attention_mask = model_input != self.null_idx
+                model_input = torch.cat([encoder_state, input], dim=-1)
+            attention_mask = model_input != self.NULL_IDX
+            position_ids = (
```
Comment (emilydinan) [inline parlai/agents/hugging_face/gpt2.py:92]: nice

**Review by emilydinan — APPROVED** [review]
woof this was a doozy but lgtm

File: parlai/agents/hugging_face/gpt2.py (hunk 1 above)
Comment (stephenroller) [reply parlai/agents/hugging_face/gpt2.py:48]: Add start token just says whether we use the start token or not, not whether we add it.

File: parlai/agents/hugging_face/gpt2.py [hunk 5]
```diff
@@ -35,12 +34,21 @@ def __init__(self, opt, dict):
...
         self.transformer = self._init_from_pretrained(opt)
         # add special tokens
-        self.start_idx = dict.start_idx
-        self.null_idx = dict.null_idx
-        self.add_start_token = False
         if opt["add_special_tokens"]:
+            size_before = self.transformer.wte.weight.size(0)
             self.transformer.resize_token_embeddings(len(dict.tokenizer))
-            self.add_start_token = opt["add_start_token"]
+            with torch.no_grad():
```
Comment (chinnadhurai) [inline parlai/agents/hugging_face/gpt2.py:40]: Wondering as to why we need this ?

**Review by chinnadhurai — APPROVED** [review]

File: parlai/agents/hugging_face/gpt2.py (hunk 5 above)
Comment (stephenroller) [reply parlai/agents/hugging_face/gpt2.py:40]: The torch nograd or the initialization changes?
